### ISSUE TITLE
[BUG] - Increased memory and cpu usage on 1.24.2 because of "HardForkEncoderDisabledEra"

### ISSUE THREAD

**Issue by gitmachtl** [input-output-hk/cardano-node#2235]
**External**

**Summary**
It looks like many relay nodes of SPOs are showing an increase memory usage and memory leaking combined with an increased cpu usage on the Allegra MainNet with 1.24.2 caused by other nodes that have not forked yet to Allegra. Theses nodes are producing "HardForkEncoderDisabledEra" Error messages in the logfiles of the relay nodes.

**Expected behavior**
No influence of none forked nodes on the performance of the local node at all. Cardano-Node should be immutable agains such kind of things (wrong eras, wrong magic-params, ...)

**System info (please complete the following information):**
- OS: Generic Linux, Debian, Ubuntu
- Node version: 1.24.2

**Additional context**
My relay nodes used around 5GB of ram after 12hrs of uptime, i have now introduced a temporary solution using fail2ban to block such node ips from the relay node for one day. After implementing that, i now see memory usage of around 1,8GB after 11 hrs of uptime. Instructions for the fail2ban filter can be found here: https://github.com/gitmachtl/scripts/blob/master/BlockNodesViaFail2Ban.md

I am not 100% sure at the moment if this is the issue causing the memory leak and higher cpu usage, but it helped SPOs so far from what i have heard back.

**Comment by drastys** [comment]
Hello,
similar for me, my node consumed 7GB RAM for first time in MN. [DOLCA]

System info (please complete the following information):
    OS: Generic Linux, Debian, Ubuntu
    Node version: 1.24.2

**Comment by Straightpool** [comment]
I can confirm the issue on 1.24.2 and that the fail2ban fix helps:

First restart is without the fail2ban patch, second with:
![image](https://user-images.githubusercontent.com/42584250/102710899-088ef600-42b6-11eb-9a63-0983cfb44b55.png)

**Comment by laplasz** [comment]
even I got an out of memory error once all the memory was allocated
```
cardano-node: internal error: Unable to commit 1048576 bytes of memory
```

**Comment by rphair** [comment]
This script looks like a great option (though our own pool would have to rewrite it for `SimpleView`) but there is the separate question: whether allowing connections from the obsolete nodes serves a purpose.  I don't see why `cardano-node 1.24.2` would allow these connections, and use memory in response to them, without any purpose.  Perhaps some information still comes from the partially compliant nodes that serves the relay and/or the network?

In any case we definitely need some insight from a developer involved in this question before deciding whether to ban the old nodes: https://forum.cardano.org/t/bad-transactions-effect-on-node-performance/44428/3

**Comment by deepfire** [comment]
@gitmachtl, @drastys, @Straightpool, @rphair, can you tell how many:

1. peers
2. active TCP connections

..did the affected nodes have _at the time of peak usage_?

**Comment by rphair** [comment]
We'll wait for the problem to emerge again (should be in about 1 day) & post that count from `ss -t`, but we're still unable to get the count of peers as measured by the node itself, as explained in this issue (roughly, since we don't see this metric with ` "ViewMode": "SimpleView"` and `"FetchBlockPeerDecisions": true` even with `"minSeverity": "Info"`): https://github.com/input-output-hk/cardano-node/issues/1703

If someone can post `config.json` settings that produce a `cardano_node_BlockFetchDecision_peers_connectedPeers_int` metric from Prometheus (or tell us how to get it via EKG), or another "peers" metric to use & how we can get it, then we will include that as well.

**Comment by gitmachtl** [comment]
@rphair make sure you have
`"TraceBlockFetchDecisions": true,`

set in your config and also the backends similar to:
```
  "options": {
    "mapBackends": {
      "cardano.node-metrics": [
        "EKGViewBK",
        {
          "kind": "UserDefinedBK",
          "name": "LiveViewBackend"
        }
      ],
      "cardano.node.BlockFetchDecision.peers": [
        "EKGViewBK",
        {
          "kind": "UserDefinedBK",
          "name": "LiveViewBackend"
        }
      ],
      "cardano.node.ChainDB.metrics": [
        "EKGViewBK",
        {
          "kind": "UserDefinedBK",
          "name": "LiveViewBackend"
        }
      ],
      "cardano.node.resources": [
        "EKGViewBK"
      ],
      "cardano.node.Forge.metrics": [
        "EKGViewBK"
      ],
      "cardano.node.metrics": [
        "EKGViewBK",
        {
          "kind": "UserDefinedBK",
          "name": "LiveViewBackend"
        }
      ]
    },

```

With that i can get the `cardano_node_BlockFetchDecision_peers_connectedPeers_int` value without any issue.

**Comment by rphair** [comment]
thanks @gitmachtl - we have no use for the `LiveViewBackend` but once we added `"TraceBlockFetchDecisions": true` we got that metric. 😍

Though @deepfire I don't think that will help us in this matter because noted in the forum link at https://github.com/input-output-hk/cardano-node/issues/2235#issuecomment-757499882 our two relays, only one of which has this problem, have the same constant number of outgoing connected peers. 🤔

**Comment by deepfire** [comment]
@rphair, we have one TCP connection per peer (up- or downstream), so knowing the total number of TCP connections and the number of upstreams will give us the downstreams.

**Comment by deepfire** [comment]
Also, I'd use the following:
```
$ sudo ss -tp | grep '"cardano-node"'
ESTAB      0      0          10.20.0.9:35809      3.0.229.88:origo-native users:(("cardano-node",pid=22034,fd=39))               
```

**Comment by deepfire** [comment]
Also, to get the TCP connection count I'd use the following:
```
$ sudo ss -tp | grep '"cardano-node"'
ESTAB      0      0          111.222.0.1:35809      111.0.222.11:origo-native users:(("cardano-node",pid=22034,fd=39))               
```

**Comment by rphair** [comment]
Yes thanks @deepfire, I understood that, but some SPOs might not be able to get the Prometheus metric to work but can still report the output from `ss` or `netstat` in which the incoming & outgoing peers, as you know, will appear differently.  Piping the output through `sort` will separate them into two groups, where the lines can be counted pretty easily:

`# ss -tp | grep '"cardano-node"' | sort`

To rephrase what I said in my last comment, the number of outgoing connections will _never_ be different between our two relays: one of which _inevitably_ develops this problem, and the other of which _never_ has it.  So the nature of our outgoing connections might be causing the problem, but not their number.

**Comment by rphair** [comment]
Also I've asked (on the Cardano Forum) anyone else seeing this problem to follow your instructions about reporting on the incoming & outgoing peers as requested: https://forum.cardano.org/t/bad-transactions-or-outdated-nodes-effect-on-performance/44428/6

**Comment by rphair** [comment]
In case we have to reboot our affected node after sending only the 2 statistics requested above, and in case that shows the problem is not strictly related to the *number* of peers but some other aspect of network traffic that we can plan in advance to measure before rebooting, here's our network traffic on the last 6 hours of the "normal" vs. the "stressed" relay.

**"normal" relay (NYC): inbound traffic = outbound traffic = constant**
![nyc](https://user-images.githubusercontent.com/2572798/104435712-b38f8880-55b2-11eb-863b-21b4b5e9bd46.png)
(Note the "private" network traffic is the small bandwidth going to & from our block producer, which is on the same network segment in the data centre.)

**"stressed" relay (Singapore): inbound traffic << outbound traffic (by about factor x5) = bursty**
![sgp](https://user-images.githubusercontent.com/2572798/104435735-bbe7c380-55b2-11eb-90d2-28e140bc6930.png)

As said on the forum, these two nodes have an identical number of outgoing peers.  Interestingly the "normal" relay has twice the number of _incoming_ peers as the "stressed" relay; not the other way around.  And also as noted on the forum, the "stressed" relay is getting 4x the number of `HardForkEncoderDisabledEra` error messages in the `SimpleView` logfile.

Since our graphs go back 14 days, we can see that during the times when the "stress" was most apparent... memory use increasing so much that we saw swapping... there was no change in the amount of network traffic (we still have yet to test the number of peers as requested).  So either the network traffic is not related to the memory use climbing _or_ the memory problem is the cumulative result of some network problem which might therefore can be measured _before_ the time of peak usage.

**Comment by rphair** [comment]
We were just getting the problem on our affected relay when I posted the stats above.  The memory leak has been cutting into our swap space for about 24 hours now.  As indicated above & in our topology file, we have **9 outgoing** peers, and still only **16 incoming** peers: not really out of the ordinary nor enough to cause a memory leak on its own (since our unaffected relay usually has more than that).

According to our leadership schedule we are done producing our blocks for this epoch.  So until shortly before the end of this epoch (midnight where we are in Asia) I'll leave this affected relay up in case there is any other information we can gather.  As you know we are taking a bit of a risk by doing this but we really need to see this solved... for our own pool's security and that of the community... I doubt we want to be launching the multi-asset protocol with this bug still kicking around.

For removal of doubt, below is the output from `ss -tp | grep '"cardano-node"' | sort | sed -e 's/ *$//'` on the affected machine (with BP obscured).  We'll plan to reboot it in 36 hours so please let's get some real diagnostics done before then. 😎

```
ESTAB   0        0           128.199.113.85:34631        13.235.216.101:4321     users:(("cardano-node",pid=973658,fd=30))
ESTAB   0        0           128.199.113.85:35265         52.59.230.169:4321     users:(("cardano-node",pid=973658,fd=34))
ESTAB   0        0           128.199.113.85:38093         13.76.102.234:2025     users:(("cardano-node",pid=973658,fd=35))
ESTAB   0        0           128.199.113.85:39935          54.238.92.29:3001     users:(("cardano-node",pid=973658,fd=26))
ESTAB   0        0           128.199.113.85:43029           3.237.5.103:4321     users:(("cardano-node",pid=973658,fd=39))
ESTAB   0        0           128.199.113.85:43985          64.227.16.35:5250     users:(("cardano-node",pid=973658,fd=33))
ESTAB   0        0           128.199.113.85:44155           3.114.7.159:4321     users:(("cardano-node",pid=973658,fd=36))
ESTAB   0        0           128.199.113.85:44899           xx.xx.xx.xx:NNNN     users:(("cardano-node",pid=973658,fd=40))
ESTAB   0        0           128.199.113.85:46831         52.74.189.190:3001     users:(("cardano-node",pid=973658,fd=38))
ESTAB   0        0           128.199.113.85:5250             95.89.8.74:45029    users:(("cardano-node",pid=973658,fd=47))
ESTAB   0        0           128.199.113.85:5250            2.56.99.230:2530     users:(("cardano-node",pid=973658,fd=50))
ESTAB   0        0           128.199.113.85:5250           159.65.73.84:43135    users:(("cardano-node",pid=973658,fd=42))
ESTAB   0        0           128.199.113.85:5250           64.227.16.35:39859    users:(("cardano-node",pid=973658,fd=27))
ESTAB   0        0           128.199.113.85:5250          104.237.2.151:33879    users:(("cardano-node",pid=973658,fd=31))
ESTAB   0        0           128.199.113.85:5250          42.118.154.57:34641    users:(("cardano-node",pid=973658,fd=32))
ESTAB   0        0           128.199.113.85:5250          54.151.129.42:45577    users:(("cardano-node",pid=973658,fd=51))
ESTAB   0        0           128.199.113.85:5250          76.81.132.186:39037    users:(("cardano-node",pid=973658,fd=41))
ESTAB   0        0           128.199.113.85:5250          81.225.71.128:37103    users:(("cardano-node",pid=973658,fd=29))
ESTAB   0        0           128.199.113.85:5250         104.238.146.49:34359    users:(("cardano-node",pid=973658,fd=28))
ESTAB   0        0           128.199.113.85:5250            xx.xx.xx.xx:NNNNN    users:(("cardano-node",pid=973658,fd=45))
ESTAB   0        0           128.199.113.85:5250         173.249.41.117:35909    users:(("cardano-node",pid=973658,fd=43))
ESTAB   0        0           128.199.113.85:5250         209.126.12.123:33285    users:(("cardano-node",pid=973658,fd=46))
ESTAB   0        0           128.199.113.85:5250        157.245.219.201:41635    users:(("cardano-node",pid=973658,fd=53))
ESTAB   0        0           128.199.113.85:5250        207.180.253.140:11941    users:(("cardano-node",pid=973658,fd=37))
ESTAB   0        1068        128.199.113.85:5250          76.81.132.186:32947    users:(("cardano-node",pid=973658,fd=52))
ESTAB   0        1068        128.199.113.85:5250          76.81.132.186:37373    users:(("cardano-node",pid=973658,fd=58))
```

**Comment by karknu** [comment]
@rphair It would be helpfull if you could provide a pcap of the data for the clients that connect to you. I will try to reach out to you on telegram.

**Comment by rphair** [comment]
for other observers... just sent @karknu a traffic sample (pcap) from the adversely affected machine.

**Comment by rphair** [comment]
A Telegram conversation with @karknu about the linked issue above confirmed there is a shrinking number of these bad (obsolete node version) peers.  We've decided to wait for that fix rather than installing the ban script.

One reason is that we use `SimpleView` text logging rather than JSON, but it's still possible to get the bad peer IPs.  This can be done by hand since we've seen very few of these all along... going back to just before the Allegra upgrade:

```
$ nice grep HardForkEncoderDisabledEra node.log | sed 's/.*String "ErrorPolicyTrace",String "//' | sed 's/:.*//' | sort -u
13.95.154.192
172.105.225.197
185.169.254.250
195.201.112.148
61.75.68.87
64.227.32.76
96.126.103.227
```

And if we check just the last 10000 reports of that error message (for us, going back a day or two), we have only one non-compliant node left which is apparently causing all of our current trouble:

```
$ nice grep HardForkEncoderDisabledEra node.log | tail -n10000 | sed 's/.*String "ErrorPolicyTrace",String "//' | tail | sed 's/:.*//' | sort -u
64.227.32.76
```

... so just banning that one IP address would be enough for us.

I just have one question for others, which I'm having trouble with: since our data centre firewall has no way to ban a single IP address, we'd have to also enable UFW as an interior firewall, with everything enabled by default but just one IP address denied.  Yet I can't find a UFW rule to "allow everything from everywhere."  Setting that wide-open UFW default plus the single ban rule would be a very small-footprint solution for this.  If anyone has an answer to that, please post here. 🙏

**Comment by gitmachtl** [comment]
The problem is, we have to get to the bottom of the problem within cardano-node itself. Otherwise we will see the same this when we fork to mary in a few weeks. For now, keeping the baning in place does not hurt.

**Comment by rphair** [comment]
The devs are working on it for sure based on what I saw today.  Not all the work they're doing on the linked issue is being recorded by those devs (one mentioned a separate group) so I think we have their full attention.  This will get resolved by the next fork for sure.  It sounded like further progress on that issue will be posted in https://github.com/input-output-hk/ouroboros-network/issues/2870.

BTW in case it helps anyone else in setting up a _temporary_ blacklist, the command to set a global whitelist in UFW (prior to adding any "deny" rules) is: `ufw default allow incoming`

**Comment by karknu** [comment]
With help from @rphair we've managed to reproduce the problem.

**Comment by rphair** [comment]
@karknu just heard that this has been effectively fixed in https://github.com/input-output-hk/ouroboros-network/pull/2880.  If that is not true please let us know 🙏

**Comment by karknu** [comment]
@rphair Yes the problem is fixed in ouroboros-network and once cardano-node is updated to use a newer version of ourboros-network this issue can be closed.

**Comment by gitmachtl** [comment]
Sounds good, lets check this with 1.25.0

**Comment by karknu** [comment]
Fixed with the merge of #2291.
